### PR TITLE
gh-101140: Add free list for generators and async functions

### PR DESCRIPTION
**Pull request by DinoV**
This adds a free list for generator functions, async generators, and co-routines.  On Instagram's workload a similar change has yielded a .25% - .46% perf win on the production workload.  This differs from that slightly in that since 3.10 generators are now variable sized objects.  Therefore the cache is a little bit bigger to accommodate the different sizes, but has a smaller number of entries per-size, and is arbitrarily limited at a certain size (currently 20).  These numbers could likely be tuned further.

* Issue: gh-101140

### CODE REVIEW

**Comment by DinoV** [comment]
On the difference from 3.10 we have the size set to 200.  If we tune the size there to 500 it's slightly better (.03%), if it's 20 it's a little worse (.11%) but those are both within the error margins.  So I think the slightly larger cache (as it's now up to 400) to accommodate the different sizes in `main` is probably relatively good.

**Comment by pablogsal** [comment]
@markshannon can we run the automatic benchmark on this PR?

**Comment by markshannon** [comment]
While getting more freelist hits is a good idea (the current hit rate is only [about 46%](https://github.com/faster-cpython/ideas/blob/main/stats/pystats-2023-01-22-python-c1c5882.md)) adding freelists for small ints would be be *far* more effective.
Adding a set of size-based freelists would be truly effective.

See https://github.com/faster-cpython/ideas/discussions/132 for more of my thoughts on the matter.

In short, I agree this a problem worth solving but I don't think this is the solution we want.
